python/test/unit/language/test_core.py

```diff
@@ -3329,6 +3329,50 @@ def kernel(
     np.testing.assert_allclose(out_ref, to_numpy(out_tri), rtol=0.01, atol=1e-2)
 
 
+@pytest.mark.interpreter
+def test_dot_optimize_epilogue(device):
+
+    # triton kernel
+    @triton.jit
+    def kernel(X, stride_xm, stride_xk, Y, stride_yk, stride_yn, M, N, K, BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr,
+               BLOCK_K: tl.constexpr, Out, stride_om, stride_on):
+        off_m = tl.arange(0, BLOCK_M)
+        off_n = tl.arange(0, BLOCK_N)
+        off_k = tl.arange(0, BLOCK_K)
+        Xs = X + off_m[:, None] * stride_xm + off_k[None, :] * stride_xk
+        Ys = Y + off_k[:, None] * stride_yk + off_n[None, :] * stride_yn
+        mask_x = (off_m < M)[:, None] & (off_k < K)[None, :]
+        mask_y = (off_m < K)[:, None] & (off_n < N)[None, :]
+
+        x = tl.load(Xs, mask=mask_x, other=0.0)
+        y = tl.load(Ys, mask=mask_y, other=0.0)
+
+        z = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.float32)
+        z += tl.dot(x, y)
+        Outs = Out + off_m[:, None] * stride_om + off_n[None, :] * stride_on
+        mask_o = (off_m < M)[:, None] & (off_n < N)[None, :]
+
```

Comment:
how is this different that what is in test_dot?

Reply:
In this test, BLOCK_M > M, so it's testing whether masked store is working correctly, which is not tested in test_dot.

Reply:
1. The problem is with the mask here, and kernel in `test_dot` does not use mask. 
2. Output tensor is of shape (2*M, N), the second half is to check there is no out-of-bound memory access.